### DISCUSSION TITLE
Is it possible to inject component state from a custom directive?

### DISCUSSION THREAD

**Discussion by thiagomajesk** [alpinejs/alpine#3649] [1. Help] [answered]
Hi! I've been outside of the 'alpine loop' for quite some time and I was wondering if it's possible to modify/ transform or inject state into a component from another directive... Let me explain what I'm trying to achieve:

Let's say we have a `todo` component that I want to initialize, I could pass the state as `x-data="todo([])` or through `x-init=init([])`. However, what if I want to remove the data from the HTML so it's not visible after the component is initialized? 

I'm using Elixir/ Phoenix, so an example would be something like this: 

```html
<!-- `x-from` receives the stringified values we want to use -->
<div x-data="todo" x-from={Jason.encode!(@items)}>
  <!-- Logic that deals with todo items -->
</div>
```

By being able to inject a JSON value directly into a different HTML attribute (`x-from`), I can avoid having to concatenate the component initialization with the actual data coming from the templating engine, and this would give me the opportunity to remove the `x-from` attribute from the DOM, so the information is not available to the end user. Is something like this even possible right now!?

**Comment by thiagomajesk** [comment]
Now that I'm thinking a little more about this and since this is such a common use-case while interopping Alpine with HTML, perhaps a better API would be to allow `x-init` to receive data and pass it directly to the component's `init()` function, like this: `x-init="[]"` or perhaps `x-init.with="[]"` to keep backward compatibility - or even `x-init.with.hide="[]"` so we can protect the initialization information.

**Comment by ekwoka** [comment]
You can, int he directive use the magics provided to add it to the data.

But also, the removing it part doesn't mean anything. The use can still see it, either by disabling js, or whatever.

The data still exists, and it's on their device.

It's not private.

Also, not sure what "protect the initialization" means.


The standard way is to pass it directly to the data constructor.

A common alternative I see with how template systems work is to put it in a json script tag as the first child and parse it from there.

**Reply by thiagomajesk** [reply]
Not sure I understand your comment completely… What exactly would be your solution here?

> Also, not sure what "protect the initialization" means.

Of course the data exists in the user device, but it would require users a little more than inspecting the page to see it. Besides that, it’s helpful to not have the information polluting the DOM.

**Reply by ekwoka** [reply]
For whom? Who is looking at the Dom? What does it matter if the person inspects and sees the data?

But yes, you can use the utilities giving to directives to change and add data to a component context.

**Reply by thiagomajesk** [reply]
> For whom? Who is looking at the Dom? What does it matter if the person inspects and sees the data?

Perhaps I wasn't clear on the original question, but I'm more interested in the "how" and not the "why" - thanks for the convo though. I'll take a closer look at the documentation to see what's available. In the meantime, I'll leave the discussion open if other people want to contribute. Cheers!

**Reply by SimoTod** [reply]
The "why" is quite important though or you will end up doing something really complicated for no reason.
In theory, you can have your own evaluator, automatically decrypting stuff.
Or you can have your custom directive acting likr x-data and decrypting thing when initialising. 
If your concern is just about having to much data in the html, just use alpine.data and keep your fields in a script tag.

**Reply by thiagomajesk** [reply]
Thanks for your input @SimoTod...

> The "why" is quite important though or you will end up doing something really complicated for no reason.

I'm not saying the "why" is not important, I'm just saying that I'm *more* concerned with the "how" (so we don't get beyond the scope of the question). Besides that, my intention is not to get the discussion sidetracked to be about the reasons this might be relevant or not 😅.

> In theory, you can have your own evaluator, automatically decrypting stuff.

Do I have a way to guarantee that this custom directive runs immediately after (or perhaps before) `x-init`? I think that the standard way of accessing data inside the component scope is using `$data`, but I see there are other (perhaps helpful) functions like `Alpine.addScopeToNode` that are not documented, is there something else I should know?

**Reply by SimoTod** [reply]
There's a lot of undocumented feature that you could use. $data accesses the closest scope which is not necessarily x-data however, most of the time you don't need $data, the scoped variable is always available in the component. A custom directive can be created with a specific priority to run before or after another one (see https://alpinejs.dev/advanced/extending#custom-order). 
In theory you can have a custom directive that acts as an encrypted x-data which decripts the value and adds a scope to the node (obviously there will be a few stuff to do, you can check the x-data source code to see what it does)

**Reply by ekwoka** [reply]
You don't need to do your own evaluator/encryption. It's pointless.

A custom directive is given access to utilities that can access and modify the data of a component.

And yes, the docs show how to define where a directive should run in the loading order.

Could then remove itself from the Dom is that still seems important.

**Comment by docdunning** [comment]
I have effectively the same question - is it possible to access/update the current component's state from a custom directive?

I can access stores from within the directive, but I'm struggling to access the state.  I've tried `Alpine.data` and `Alpine.$data`, which both appear to be functions, but `Alpine.$data('varname')` and other variations just give me `undefined`.

@ekwoka - you said .. 

> in the directive use the magics provided to add it to the data.

But I can't work out how to do that :-(  - what's the syntax?

Here's a cut-down version of the directive

```
  Alpine.directive('uppercase', function (el, {value, modifiers, expression}, {Alpine}) {
    el.textContent = el.textContent.toUpperCase()
    const store = Alpine.store('mystore');
    store.foo = 'bar' // This works fine
    console.log (Alpine.$data('mydata')) // This returns undefined
  })
```

If I can't access the component scope, I'll just use the store, but it would be handy to know if the component scope can be accessed.

**Reply by ekwoka** [reply, marked as answer]
`Alpine.$data` accepts an element as the argument. So you can get the datastack at that element.

do to fix your uppercase

```js
  Alpine.directive('uppercase', (el, {value, modifiers, expression}, { effect }) {
    const data =Alpine.$data(el);
    effect(() => el.textContent = data[value].toUpperCase())
  })
```
  
  Or whatever

**Reply by docdunning** [reply]
> Alpine.$data accepts an element as the argument

Aha - that's working fine.  Thanks :-)

**Reply by ekwoka** [reply]
(TypeScript would tell you that :smile:)

**Comment by trych** [comment]
> (TypeScript would tell you that 😄)

Unlike the docs, I believe?

**Reply by ekwoka** [reply]
Technically it's undocumented (I do not know why the undocumented things are undocumented, especially one like this that SEEMs pretty solid and stable in its API).
